rancher/rancher
# Use GUID for the PrincipalID for Active Directory

## Issue: https://github.com/rancher/rancher/issues/37994

 
## Problem
If Rancher is configured to use AD for authentication, and AD user accounts get moved between OUs, then when the users log back into Rancher a new user is created and all their old permissions are not preserved.
This is because the principalID was based on the DN of each user.  When the OU changes, it results in a new DN, which results in a new Rancher user.
 
## Solution
Active Directory has an attribute, objectGUID, which is unique and immutable.  We now use that attribute as the basis for the principalID
 
## Testing
The repro steps in the original issue are correct.

## Engineering Testing
### Manual Testing
With a fresh Rancher install, set up Active Directory as the auth provider.  Verify that you can login as usual with valid users from that auth provider.  All Rancher functionality should appear normal.
You can verify that the user has been assigned a principalID based on GUID rather than DN by doing:
`kubectl get user <user> --template='{{.principalIds}}'`
The output of that should show one of the principalIds that will look like:  `activedirectory_user://bfb34c007dc2c843adcc74ac3e27df21`
instead of the old-style, DN-based principalId, which would look like:
`CN=username,CN=Users,DC=qa,DC=company,DC=org`

I've also tried changing the OU of a user and made sure that the OU change didn't result in a new Rancher user (the objectGUID mapping IS correctly working and Rancher is no longer relying on the DN)

### Automated Testing
Unit tests

## QA Testing Considerations
The upgrade scenario is important for this issue.  This PR includes a "migration" function that will run at Rancher startup time to update user, token, crtb, prtb objects that were using a DN-based solution to use the GUID as the principal.  In order to test this, you would need to either have an older version running with Rancher users created, and then upgrade.  Or, it's possible to have some sort of "reset" script that will take AD Rancher users and "reset" the principalID to be DN-based (and the corresponding objects....note, CRTBs and PRTBs will need to be deleted and re-created instead of updated in place.  I'm happy to work with QA to come up with some sort of "reset" script that will work for them.
 
### Regressions Considerations
If the migration process were to fail to update one or more bindings, the affected user would see their permissions affected negatively, so it's especially important to verify that the user's bindings were all updated as part of the migration process.
